### DISCUSSION TITLE
Gladiator on secondary road / Entertainment

### DISCUSSION THREAD

**Discussion by catalin-duta** [Keriew/augustus#521] [Q&A]
I have some confusions given the below situation.
Latest build 3.0.1.287-d5737793

![Screenshot_20211111-102007543](https://user-images.githubusercontent.com/93785920/141263175-0d3ea580-17ba-4f55-b910-b979ba71e94c.jpg)
The elevated part in the screenshot plus the small road in the north, are on a separate road network from the rest of the city.
It has one each of Arena, Theater and the 3 entertainment schools.

How they behave:
1. Gladiators never spawn from the school; the Arena has never seen one, nor has an Amphitheater that used to be in the same area. Occasionally a Gladiator will come out of the Arena (why?) but he does nothing for the houses.
2. Are the entertainment destination walkers supposed to be proactive (i.e. they head out before the show ends), or reactive (they head out when the show ends), or downright late? I'm asking because:
2.1. Actors leave from the Colony more or less when the show in the Theater ends. The Colony puts out 3 or 4 of them, roughly 15 tiles behind one another. These guys seem to be reactive, but not very smart. There's a walker already on its way, but they keep coming.
2.2. Lion tamers leave the Lion Pit some 1-2 months after the show in the Arena has ended. There's generally 2 of them on the way, some 20 tiles apart. Are they meant to be so late?

Savegame also attached.
[6Lugdunum8Ah19.zip](https://github.com/Keriew/augustus/files/7518943/6Lugdunum8Ah19.zip)

**Comment by crudelios** [comment]
> Gladiators never spawn from the school; the Arena has never seen one, nor has an Amphitheater that used to be in the same area. Occasionally a Gladiator will come out of the Arena (why?) but he does nothing for the houses.

This may be a bug? I'm not sure, I'll need to check.

> Are the entertainment destination walkers supposed to be proactive (i.e. they head out before the show ends), or reactive (they head out when the show ends), or downright late? I'm asking because:

The only entertainment destination walkers that exist are the ones going from the school to the venue. The entertainers spawned by the venues are roaming walkers instead. They should head out while there is a show going on and stop going out when the show ends, except for theaters, which will always spawn actors provided they are staffed.

**Reply by catalin-duta** [reply]
To clarify the last part, regarding destination walkers going from school to venue.
When are they supposed to be spawned? I presume shortly after a show has ended and they are needed, because I don't think they are smart enough to leave in advance (that's an idea), so there always a show.
If they're supposed to spawn after the show ended, Lion tamers are very late.

Also, are they in any way scheduling where to go? In the case with the Actor Colony, if there were two Theaters without shows, will they spawn 3-4 times towards Theater №1 and only after they reach it, start heading to №2?

**Reply by crudelios** [reply]
Actor colonies and gladiator schools buildings can spawn walkers about six times per in-game month.
Lion pits spawn lions about 3 times per month.
Chariot makers spawn chariots at least twice per month.

Note that these spawn rates assume full entertainer school employment. Entertainers are spawned at a much slower rate if you don't have full employment for the building, and how slowly they spawn will depend on the actual number of workers you have. As a rule of thumb, even one fewer worker than needed will make spawning twice as slow.

The spawned entertainers decide which building to go to based on a heuristic value, which is calculated based on the number of days of show remaining and the distance to the venue. It will always pick the closest of this value.

Internally, the game loops through all possible destination buildings that are on the same road network and performs the following calculation:

`result = 2 * <days of show left> + <linear distance to venue>`

If the `result` of the current building is smaller than the stored result, the current building ID will be stored as the best destination so far.

When the loop through the buildings ends, whatever building ID is stored will become the actual destination.

Therefore, it may happen that if you have a venue that is too far away, a closer venue will still be selected even if it just had a walker go to it.

> Also, are they in any way scheduling where to go? In the case with the Actor Colony, if there were two Theaters without shows, will they spawn 3-4 times towards Theater №1 and only after they reach it, start heading to №2?

Yes, exacly. This can be changed though so entertainers check for a better spot every now and then while moving towards a destination. I've done it for cartpushers so I can as easily do it for entertainers.

**Reply by catalin-duta** [reply]
My experience on the above road network (the rest of the city is much harder to analyse), with full employment, walkers going from school to venue:
- Actors spawn only when there is no show in the Theater, generally right after the show has ended. The theater still has down times, because Actors stop spawning while there is a show.
- Gladiators don't spawn at all
- Lions also don't spawn during a show. And checking it periodically, it took 1-2 months for them to get the news that the Arena was out of lions and spawn...

I don't know what to say about the walkers changing destination mid-trip. If they follow the rule you said and leave the school in time for the show not to end, that should not be needed.

**Reply by Keriew** [reply]
There was an oversight with how entertainment schools chose their destinations, should be better with 4c9c8ef (but it could still use some improvement).

**Comment by crudelios** [comment]
@catalin-duta Can you test the new behaviour?

**Reply by catalin-duta** [reply]
Destination walkers (from school to venue) are now constantly spawned and the Arena is fully manned.
I added another Arena at around double the distance from the school and, while they clearly prefer the closer one, they will head for the one further away as well, some 25 days before the show ends (I haven't checked with the formula). The furthest arena has some downtime, to be expected because it's so far away.
The actors are also nicely heading out towards different theaters.
All in all, destination walkers very good.
The only possible improvement I could mention is that a school try to pick the second-best destination formula-wise if it already has a walker on the way to the first-best.

There's still something I'm unsure about: is the Arena meant to spawn roaming gladiators as well, like it does lions?
The two Arenas do spawn lions, but there was only 1 roaming gladiator spawned, right after I loaded the new build and before there was a show. Ever since there are shows, there are no more roaming gladiators.
